### PR TITLE
971 release tag seems wrong

### PR DESCRIPTION
**Pull request by BPerlakiH**
Fixes: #971

I've updated the Readme file and the CD workflow to have our new approach documented and enforced.

### CODE REVIEW

**Review by kelson42 — CHANGES_REQUESTED** [review]
@BPerlakiH I believe this was still not clear to you how I want it to work.

I have fixed the `README.md` but:
* The section "Last Step" is probably still wrong
* The worflow is still wrong for the release.

There is no something like a "post release step", this is one release process which is triggered by the GitHub release and then go to testflight. It's not testflight and then github release.

**Comment by BPerlakiH** [comment]
Ohh, it is still not clear for me...
This is my understanding of it:
We need to build and upload a binary build for Apple to review.
We do this via CD (either weekly or with the TestFlight tag).
For the AppStore review we select (one of the) binaries already uploaded (usually the latest).
Once the binary is approved, we cannot change it to something else, eg. triggering the Github release -> build new binary -> upload to App Store Connect.
So my understanding is that once the binary (previously built) is approved by Apple, we need to tag it on Github as a release. This would be sufficient enough, but we want a dmg file as well. So we rebuild it from the same commit but only for FTP .dmg file.

**Comment by kelson42** [comment]
> For the AppStore review we select (one of the) binaries already uploaded (usually the latest).

No, this has to be one we have released (so we a a clear release number).

**Comment by BPerlakiH** [comment]
> > For the AppStore review we select (one of the) binaries already uploaded (usually the latest).
> 
> No, this has to be one we have released through the usual process. The validation of Apple does not decide what we release. We decide what we release and Apple decides if it goes through to the audience of its App store.

Ok, I see your point now. This is how I understand it:
1) we can create testflight builds as many we like, it won't affect either the public release to the app store or the GitHub release, or the FTP release
2) for each App Store release, we trigger an appropriate Github release, so for example:
- we Github release 3.6.0 that creates a github tag, builds everything and uploads it to Apple, Apple reviews it the next day or so
- if 3.6.0 is rejected 
- we change some code
- create Github release 3.6.1, that builds everything uploads it to Apple, Apple reviews it, and approves the next day
- we press the public release button on App Store Connect, so it's available for everyone publicly
- at this point, we still need a "post App Store release step" to create and upload the .dmg binary to FTP, to make that publicly available and the very same that is available in the App Store.

**Comment by BPerlakiH** [comment]
@kelson42 After re-reading your changes I think we are missing some pieces, to get a common understanding of this process.

We can create a release version, eg: 3.6.0. This is stored in git within the code. As time passes we weekly / or on demand create builds for this, that is available to try out. For clarity let's call them release candidates. This possibility of trying them out is available via TestFlight. The idea is that it should be tested on real devices before the public sees it.
Under each release version, we have several builds, eg this is for 3.6.0:

Screenshot 2024-10-20 at 00 59 18

Those builds are uploaded there via our TestFlight CD process. Any of those builds can be picked for a version to be released. This is what I meant that we usually pick the latest one.

So far we were following the safest path, that the build (release candidate) we try out is the one we send to Apple. If it's approved, we want to mark it as a "final" release, and as such it should be reflected in GitHub release and as a tag as well. At the same time, we do want to upload this to FTP so it's available there as well. These 2 last steps is automated, by the formerly called "post Release step": 
1) "visible" as a Github release
2) uploaded to FTP
maybe we should call it 'post App Store step'.

From all of this, what we haven't really practiced so far is creating the proper GitHub release tag, and created too many "testflight" tags with various suffixes.


Now we can turn this around, and do a new build and tag as a GitHub release, which will produce a new build on that list, and retest that, and pick it to be published to the App Store. If all is good, it still requires us to do the FTP step, once it's approved.
With this approach though we might end up with more GitHub release versions, than App Store versions, since we will GitHub release and tag even the ones that were not approved. 

Eg: version 1.6.0 is not approved by Apple:

| App Store:| GitHub:|
|-------|-------|
| 1.6.1 | 1.6.1 |
| 1.5.0 | 1.6.0 |
|       | 1.5.0 |

**Comment by BPerlakiH** [comment]
In this whole Apple process of getting from code to public App Store, the release version is fixed from the very beginning. The build number varies though.
We can go against that, and say if Apple rejects a specific version we start all over and create a new one with a patch number higher, that's OK with me. I would still prefer to have matching App Store released versions with GitHub released versions.

**Comment by kelson42** [comment]
> So that the "release" means it is publicly available both on App Store

We primarely release a code (version), all the rest is secondary.

**Review by rgaudin — APPROVED** [review]
LGTM

**Review by rgaudin — APPROVED** [review]
👍

**Comment by kelson42** [comment]
I don't know if this is ready to review but at a first look I don't see how the release event would here push on testflight.

File: README.md [hunk 1]
```diff
@@ -146,25 +146,50 @@ https://developer.apple.com/documentation/safari-developer-tools/inspecting-ios
...
+Once satisfied with the quality of the app in TestFlight, we can
+proceed with the Kiwix release process.
+
+The release process is triggered by a [GitHub
+Release](https://github.com/kiwix/kiwix-apple/releases). Only the
+version released (to Testflight) that way can be then sent to
+approval to Apple. Once approved by Apple, we can release them to the
+AppStore. At the same time, will publish the macOS DMG to our [file
```
Comment (rgaudin) [inline README.md:181]: This *At the same time* is confusing. It should be clear that this is triggered by the Github release event.

**Comment by rgaudin** [comment]
> I don't know if this is ready to review but at a first look I don't see how the release event would here push on testflight.... Or it was already doing it in the past?

Indeed, I thought this was a doc-only PR. What is described in this PR's readme is not what is happening in the workflow. We are setting UPLOAD_TO differently. This will require refactoring of the workflow as it was previously either one or the other

**Comment by kelson42** [comment]
@BPerlakiH Can you please focus on fixing this? This is a very old PR and I want this working fine before releasing.

File: README.md (hunk 1 above)
Comment (BPerlakiH) [reply README.md]: Thanks @rgaudin. I agree, please see it updated.

File: .github/workflows/cd.yml [hunk 2]
```diff
@@ -50,6 +52,10 @@ jobs:
     - name: Install python dependencies
       run: pip install pyyaml==6.0.1
 
+    - name: Remove Apple Pay capability for macOS
```
Comment (BPerlakiH) [inline .github/workflows/cd.yml:55]: This is to fix: #753 builds

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -41,6 +41,10 @@ jobs:
         if: matrix.platform == 'iOS'
         run: echo "EXTRA_XCODEBUILD=-sdk iphoneos ${{ env.APPLE_AUTH_PARAMS }}" >> $GITHUB_ENV
```
Comment (BPerlakiH) [inline .github/workflows/ci.yml:43]: Also to fix: #753 builds

**Review by rgaudin — DISMISSED** [review]
LGTM but I find it a tad less readable.

I'm not enjoying the use of bash array for `UPLOAD_TO`. We've been using the GITHUB_ENV solely as a mean to store data between steps. It was dumb on purpose.

I dont know if the [`contains()`](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#contains) expression is matching on it as an array or as strings.

@BPerlakiH could you please carify? If this matches as string, then we can simplify the syntax and use `UPLOAD_TO="app-store ftp"` instead of `UPLOAD_TO=("app-store" "ftp")`.

**Comment by BPerlakiH** [comment]
> LGTM but I find it a tad less readable.
> 
> I'm not enjoying the use of bash array for `UPLOAD_TO`. We've been using the GITHUB_ENV solely as a mean to store data between steps. It was dumb on purpose.
> 
> I dont know if the [`contains()`](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#contains) expression is matching on it as an array or as strings.
> 
> @BPerlakiH could you please carify? If this matches as string, then we can simplify the syntax and use `UPLOAD_TO="app-store ftp"` instead of `UPLOAD_TO=("app-store" "ftp")`.

I also think it's less readable than it was before... This was the least destructive way to change it. We have both EXPORT_METHOD and the Signing Certificate different when it comes to macOS ftp vs macOS App-Store, so without changing the XCodeBuild step as well we need to run those in "parallel".

I checked the ``contains`` method, and it is in fact does string matching, so I changed it accordingly.

**Review by rgaudin — APPROVED** [review]
All good now

**Pull request merged**